monai/metrics/mmd.py

```diff
@@ -0,0 +1,87 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from collections.abc import Callable
+
+import torch
+
+from monai.metrics.metric import Metric
+
+
+class MMDMetric(Metric):
+    """
+    Unbiased Maximum Mean Discrepancy (MMD) is a kernel-based method for measuring the similarity between two
+    distributions. It is a non-negative metric where a smaller value indicates a closer match between the two
+    distributions.
+
+    Gretton, A., et al,, 2012.  A kernel two-sample test. The Journal of Machine Learning Research, 13(1), pp.723-773.
+
+    Args:
+        y_transform: Callable to transform the y tensor before computing the metric. It is usually a Gaussian or Laplace
+            filter, but it can be any function that takes a tensor as input and returns a tensor as output such as a
+            feature extractor or an Identity function.
+        y_pred_transform: Callable to transform the y_pred tensor before computing the metric.
+    """
+
+    def __init__(self, y_transform: Callable | None = None, y_pred_transform: Callable | None = None) -> None:
+        super().__init__()
+
+        self.y_transform = y_transform
+        self.y_pred_transform = y_pred_transform
+
+    def __call__(self, y: torch.Tensor, y_pred: torch.Tensor) -> torch.Tensor:
+        return compute_mmd(y, y_pred, self.y_transform, self.y_pred_transform)
+
+
+def compute_mmd(
+    y: torch.Tensor, y_pred: torch.Tensor, y_transform: Callable | None, y_pred_transform: Callable | None
+) -> torch.Tensor:
+    """
+    Args:
+        y: first sample (e.g., the reference image). Its shape is (B,C,W,H) for 2D data and (B,C,W,H,D) for 3D.
+        y_pred: second sample (e.g., the reconstructed image). It has similar shape as y.
+    """
+
+    # Beta and Gamma are not calculated since torch.mean is used at return
+    beta = 1.0
+    gamma = 2.0
+
+    if y_transform is not None:
+        y = y_transform(y)
+
+    if y_pred_transform is not None:
+        y_pred = y_pred_transform(y_pred)
+
+    if y_pred.shape != y.shape:
+        raise ValueError(
+            "y_pred and y shapes dont match after being processed "
+            f"by their transforms, received y_pred: {y_pred.shape} and y: {y.shape}"
+        )
+
+    for d in range(len(y.shape) - 1, 1, -1):
+        y = y.squeeze(dim=d)
+        y_pred = y_pred.squeeze(dim=d)
+
+    y = y.view(y.shape[0], -1)
+    y_pred = y_pred.view(y_pred.shape[0], -1)
+
+    y_y = torch.mm(y, y.t())
+    y_pred_y_pred = torch.mm(y_pred, y_pred.t())
+    y_pred_y = torch.mm(y_pred, y.t())
+
+    y_y = y_y / y.shape[1]
+    y_pred_y_pred = y_pred_y_pred / y.shape[1]
+    y_pred_y = y_pred_y / y.shape[1]
+
+    # Ref. 1 Eq. 3 (found under Lemma 6)
```

Comment:
The formula used here is different than the Lemma 6 in Ref. 1
```math
uMMD^2 = \beta (\text{mean}(y \cdot y) + \text{mean}(y_{pred} \cdot y_{pred})) - \gamma \text{mean}(y_{pred} \cdot y)
```
Would it be possible to explain the cause of the discrepancy? Thanks

Reply:
Hmm you're right, comparing to equation 3:

<img width="761" alt="image" src="https://github.com/Project-MONAI/MONAI/assets/7947315/8fbe8dbf-83e4-413d-a50b-4ff8badb884b">

Here $k(x_i,x_j)$ gives the same result as $torch.mm(y, y.t())$ - a matrix we can index using $i,j$, if the kernel function $k$ takes the inner product between $x_i,x_j$. Then given $\beta$ and $\gamma$ are fixed to 1 and 2 in the code, I'm not sure what lines 82-84 are doing. We also need to exclude the diagonal components of e.g. $y_y$ from the calculations.

I didn't write this code myself, going to contact the author and ask why it is different.

Reply:
Thank you for contacting the author @marksgraham . The formula looks like some approximation to me. Feel free to skip and close this one when you see fit, as I don't want this to block your PR.

Reply:
@marksgraham @mingxin-zheng When I implemented this I based it on an open-source code available [here](https://github.com/cyclomon/3dbraingen/blob/master/Test.ipynb).

From what I remember lines 82-84 are there to account for different dimensionalities `(number of columns)` of the kernel `(y_transform/y_pred_transform)` before meaning which should be more a mean of the population `(number of rows)`. 

I remember discussing this at some point with @mjorgecardoso when I was getting peculiarly high numbers.

@marksgraham you are right that I did not get the right of the main diagonal. This could quickly be done similarly with [this one](https://jejjohnson.github.io/research_journal/appendix/similarity/mmd/). By grabbing the main diagonal and then creating a diagonal matrix that we subtract from the matrix of interest `(y_y, y_pred_y_pred, y_pred_y)`.

Also, I think the `torch.mean()` is wrong as it would mean dividing by `m*m` instead of `m*(m-1)` as shown in equation 3.   

Please let me know if I can be of more help.

Reply:
I've rewritten to fully match the equation being referenced. I think we should stay faithful to the original MMD implementation. I've broken it down into parts that more clearly match the calculation so it should be easier to follow.

> From what I remember lines 82-84 are there to account for different dimensionalities (number of columns) of the kernel (y_transform/y_pred_transform) before meaning which should be more a mean of the population (number of rows).

I suggest we revisit providing this as an option later, if needs be.

Reply:
If @mingxin-zheng / @danieltudosiu could can a look to make sure you're in agreement with the changes would be great!

Reply:
Hi @marksgraham , in the update, I believe `m` equals to batch_size of `y` tensors. When the batch_size is 1, it will be a zero-division error. Is this expected?